lib/async_hooks.js

```diff
@@ -255,23 +255,21 @@ class AsyncLocalStorage {
     resource[this.kResourceStore] = store;
   }
 
-  _exit() {
-    const resource = executionAsyncResource();
-    if (resource) {
-      resource[this.kResourceStore] = undefined;
-    }
-  }
-
   runSyncAndReturn(store, callback, ...args) {
+    const resource = executionAsyncResource();
+    const outerStore = resource[this.kResourceStore];
     this._enter(store);
     try {
       return callback(...args);
     } finally {
-      this._exit();
+      resource[this.kResourceStore] = outerStore;
```

Comment:
Should we take in account when the following happen:
`resource[this.kResourceStore] = outerStore;` but before the call to `_enter(...)` `resource.hasOwnProperty(this.kResourceStore) === false`
In that case we would be adding an un-needed property to the resource. But I am not sure it has any real impact 🤔

Reply:
Previous implementation was doing `resource[this.kResourceStore] = outerStore` assignment in `_exit` method. So this PR doesn't change this behavior.

As for the impact of this assignment, I'm not sure if we need to do something with it as well.

Reply:
Make sense! thanks! :D

Reply:
Checking is probably more overhead than just assigning the `undefined`. :man_shrugging: